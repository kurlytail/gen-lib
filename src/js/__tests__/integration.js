import { execSync } from 'child_process';

describe('# integration test', () => {
    beforeEach(() => {
        execSync('rm -rf testoutput');
        execSync('mkdir testoutput');
        execSync('git init', { cwd: 'testoutput' });
        execSync('git config user.email "you@example.com"', {
            cwd: 'testoutput'
        });
        execSync('git config user.name "Your Namer"', { cwd: 'testoutput' });
        execSync('git commit --allow-empty -m "Empty commit."', {
            cwd: 'testoutput'
        });
    });

    it('## should print help options', () => {
        const output = execSync('./dist/sgen.min.js -h').toString();
        expect(output).toMatchSnapshot();
    });

    it('## should error out when downstream generator not found', () => {
        expect(() =>
            execSync(
                './dist/sgen.min.js -m src/test/fixture/map.json -d src/test/fixture/design.js -e src/test/fixture -o testoutput -g some-generator -s'
            )
        ).toThrow();
    });

    it('## should generate design', () => {
        let output = execSync(
            './dist/sgen.min.js -m src/test/fixture/map.json -d src/test/fixture/design.js -e src/test/fixture -o testoutput -s'
        ).toString();
        output = output
            .replace(
                /warn: Please cherrypick changes from master-sgen-generated from .*/,
                ''
            )
            .replace(/info: git cherry-pick .*/, '');
        expect(output).toMatchSnapshot();
    });

    it('## should generate design with merge', () => {
        let output = execSync(
            './dist/sgen.min.js -m src/test/fixture/map.json -d src/test/fixture/design.js -e src/test/fixture -o testoutput -s'
        ).toString();
        output = output
            .replace(
                /warn: Please cherrypick changes from master-sgen-generated from .*/,
                ''
            )
            .replace(/info: git cherry-pick .*/, '');
        expect(output).toMatchSnapshot();

        execSync('echo "some change" >> cella', { cwd: 'testoutput' });
        execSync('git add cella', { cwd: 'testoutput' });
        execSync('git commit -m Test', { cwd: 'testoutput' });

        output = execSync(
            './dist/sgen.min.js -m src/test/fixture/map.json -d src/test/fixture/design.js -e src/test/fixture -o testoutput -s'
        ).toString();
        output = output
            .replace(
                /warn: Please cherrypick changes from master-sgen-generated from .*/,
                ''
            )
            .replace(/info: git cherry-pick .*/, '');
        expect(output).toMatchSnapshot();

        execSync('git checkout master-sgen-generated', { cwd: 'testoutput' });
        execSync('echo "some other change" >> cellb', { cwd: 'testoutput' });
        execSync('git add cellb', { cwd: 'testoutput' });

        execSync('git commit -m Test1', { cwd: 'testoutput' });
        execSync('git checkout master', { cwd: 'testoutput' });

        output = execSync(
            './dist/sgen.min.js -m src/test/fixture/map.json -d src/test/fixture/design.js -e src/test/fixture -o testoutput -s'
        ).toString();
        output = output.replace(
            /warn: Please cherrypick changes from master-sgen-generated from .*/,
            ''
        );
        output = output.replace(/info: git cherry-pick .*/, '');
        expect(output).toMatchSnapshot();
    });
});
