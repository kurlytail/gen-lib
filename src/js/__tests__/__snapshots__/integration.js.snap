// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# integration test ## should generate design 1`] = `
"info: Initializing git repository testoutput/.git
info: loading extension src/test/fixture/ext._
info: loading extension src/test/fixture/toplevel._
info: Generating testoutput/cella from template src/test/fixture/master._
info: Generating testoutput/cellb from template src/test/fixture/master._
info: Generating testoutput/toplevel from template src/test/fixture/toplevel._
info: Committed generated files
info: Attempting to use branch master-sgen-generated
info: Creating new branch master-sgen-generated
warn: Checkout out branch master
warn: Checkout out branch master
info: sgen is done
"
`;

exports[`# integration test ## should generate design with merge 1`] = `
"info: Initializing git repository testoutput/.git
info: loading extension src/test/fixture/ext._
info: loading extension src/test/fixture/toplevel._
info: Generating testoutput/cella from template src/test/fixture/master._
info: Generating testoutput/cellb from template src/test/fixture/master._
info: Generating testoutput/toplevel from template src/test/fixture/toplevel._
info: Committed generated files
info: Attempting to use branch master-sgen-generated
info: Creating new branch master-sgen-generated
warn: Checkout out branch master
warn: Checkout out branch master
info: sgen is done
"
`;

exports[`# integration test ## should generate design with merge 2`] = `
"info: Opening git repository testoutput/.git
info: Stashing branch master
info: Attempting to use branch master-sgen-generated
warn: Checkout out branch master-sgen-generated
info: loading extension src/test/fixture/ext._
info: loading extension src/test/fixture/toplevel._
info: Generating testoutput/cella from template src/test/fixture/master._
info: Generating testoutput/cellb from template src/test/fixture/master._
info: Generating testoutput/toplevel from template src/test/fixture/toplevel._
warn: Checkout out branch master-sgen-generated
warn: Checkout out branch master-sgen-generated
warn: Nothing changed in generated files
warn: Checkout out branch master
info: sgen is done
"
`;

exports[`# integration test ## should generate design with merge 3`] = `
"info: Opening git repository testoutput/.git
info: Stashing branch master
info: Attempting to use branch master-sgen-generated
warn: Checkout out branch master-sgen-generated
info: loading extension src/test/fixture/ext._
info: loading extension src/test/fixture/toplevel._
info: Generating testoutput/cella from template src/test/fixture/master._
info: Generating testoutput/cellb from template src/test/fixture/master._
info: Generating testoutput/toplevel from template src/test/fixture/toplevel._
warn: Checkout out branch master-sgen-generated
info: Committed generated files
warn: Checkout out branch master
warn: Generator branch master-sgen-generated was modified

info: sgen is done
"
`;

exports[`# integration test ## should print help options 1`] = `
"Usage:
  node sgen.js [OPTION]

Options:
  -m, --map=ARG+       mapping files
  -d, --design=ARG+    design files
  -e, --extension=ARG+ extension directories or files
  -o, --output=ARG     output directory for generated files
      --overwrite=ARG  Overwrite modes, can be one of skip, regen, error or merge                         (default: skip)
      --forceOverwrite Override overwrite flags in the generator map with the one provided on --overwrite
  -h, --help           display this help
"
`;
